src/Microsoft.ML.AutoML/AutoMLExperiment/AutoMLExperiment.cs

```diff
@@ -0,0 +1,270 @@
+            else
+            {
+                this._serviceCollection.Add(descriptor);
+            }
+
+            return this;
+        }
+
+        public AutoMLExperiment SetPipeline(MultiModelPipeline pipeline)
+        {
+            this._settings.Pipeline = pipeline;
+            return this;
+        }
+
+        public AutoMLExperiment SetTrialRunnerFactory(ITrialRunnerFactory factory)
+        {
+            var descriptor = new ServiceDescriptor(typeof(ITrialRunnerFactory), factory);
+            if (this._serviceCollection.Contains(descriptor))
+            {
+                this._serviceCollection.Replace(descriptor);
+            }
+            else
+            {
+                this._serviceCollection.Add(descriptor);
+            }
+
+            return this;
+        }
+
+        public AutoMLExperiment SetPipeline(SweepableEstimatorPipeline pipeline)
+        {
+            var res = new MultiModelPipeline();
+            foreach (var e in pipeline.Estimators)
+            {
+                res = res.Append(e);
+            }
+
+            this.SetPipeline(res);
+
+            return this;
+        }
+
+        public AutoMLExperiment SetEvaluateMetric(BinaryClassificationMetric metric, string labelColumn = "label", string predictedColumn = "Predicted", )
+        {
+            this._settings.EvaluateMetric = new BinaryMetricSettings()
+            {
+                Metric = metric,
+                PredictedColumn = predictedColumn,
+                LabelColumn = labelColumn,
+            };
+
+            return this;
+        }
+
+        public AutoMLExperiment SetEvaluateMetric(MulticlassClassificationMetric metric, string labelColumn = "label", string predictedColumn = "Predicted")
+        {
+            this._settings.EvaluateMetric = new MultiClassMetricSettings()
+            {
+                Metric = metric,
+                PredictedColumn = predictedColumn,
+                LabelColumn = labelColumn,
+            };
+
+            return this;
+        }
+
+        public AutoMLExperiment SetEvaluateMetric(RegressionMetric metric, string labelColumn = "label", string scoreColumn = "Score")
+        {
+            this._settings.EvaluateMetric = new RegressionMetricSettings()
+            {
+                Metric = metric,
+                ScoreColumn = scoreColumn,
+                LabelColumn = labelColumn,
+            };
+
+            return this;
+        }
+
+        /// <summary>
+        /// Run experiment and return the best trial result.
+        /// </summary>
+        /// <returns></returns>
+        public Task<TrialResult> Run()
+        {
+            this.ValidateSettings();
+            var cts = new CancellationTokenSource();
+            cts.CancelAfter((int)this._settings.MaxExperimentTimeInSeconds * 1000);
+            this._settings.CancellationToken.Register(() => cts.Cancel());
+            var ct = cts.Token;
+            ct.Register(() => this._context.CancelExecution());
```

Comment:
IIRC we had some issues doing this before. Something like if we cancel the context the model is running on, then the model can't be used after we return it? Something like that. Have you tested running the model again after you return it from here?

Reply:
Let me see if I can find that spot real fast. If not then I wouldn't worry about it.

Reply:
This is what we have for a comment about it. [https://github.com/dotnet/machinelearning/blob/510f0112d4fbb4d3ee233b9ca95c83fae1f9da91/src/Microsoft.ML.AutoML/Experiment/Experiment.cs#L155](https://github.com/dotnet/machinelearning/blob/510f0112d4fbb4d3ee233b9ca95c83fae1f9da91/src/Microsoft.ML.AutoML/Experiment/Experiment.cs#L155). I believe its because it messes up using the model without saving/loading it with a new context. Or something like that anyways.

Reply:
I try a few model after cancelling and the model is still usable, so I guess it's fine to consume a completed model after cancelling context?

Reply:
Sounds good to me.